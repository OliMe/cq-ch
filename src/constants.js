export const TYPE_COMMAND = 'command';

export const TYPE_QUERY = 'query';
