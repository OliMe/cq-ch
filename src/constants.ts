export const TYPE_COMMAND = 'command';

export const TYPE_QUERY = 'query';

export const IS_BROWSER = typeof window !== 'undefined';
