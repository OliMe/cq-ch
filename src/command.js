import getTransport from './event-transport/get-transport';
import { TYPE_COMMAND } from './constants';
import { checkArguments } from './helpers/argument-checker';
import { channelCreatorCfg, commandChannelCfg } from './arguments.cfg';

/**
 * Declares a channel for sent commands in the service interface.
 * @param {Array} types Types of commands for sending.
 * @param {string} context Application context e.g. Namespace of command.
 * @return {Function} Function for sending commands to channel.
 */
export default function command (types, context): Function {
  checkArguments([types, context], 'command', channelCreatorCfg);
  return command => {
    checkArguments([command], 'commandChannel', commandChannelCfg);
    if (!types.includes(command.type)) {
      throw new TypeError('Trying to send command with type not in interface.');
    }
    command = { ...command, context };
    getTransport(TYPE_COMMAND).trigger(command.type, command);
  };
}
