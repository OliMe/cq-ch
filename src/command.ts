import getTransport from './event-transport/get-transport';
import { TYPE_COMMAND } from './constants';
import { Context, Message, Types } from './types';
import { checkChannelCreator, checkCommandChannel } from './helpers/argument-checkers';

/**
 * Declares a channel for sent commands in the service interface.
 * @param {Array} types Types of commands for sending.
 * @param {string} context Application context e.g. Namespace of command.
 * @return {Function} Function for sending commands to channel.
 */
export default function command(types: Types, context: Context) {
  checkChannelCreator(command.name, types, context);
  return async function commandChannel(command: Message<undefined>) {
    checkCommandChannel(commandChannel.name, command);
    if (!types.includes(command.type)) {
      throw new TypeError('Trying to send command with type not in interface.');
    }
    const extendedCommand = { ...command, context };
    getTransport(TYPE_COMMAND).trigger(extendedCommand.type, extendedCommand);
  };
}
