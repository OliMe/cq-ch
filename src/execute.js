// @flow
import { channelCreator, takeChannelCreator } from './helpers/channel-creators'
import { TYPE_COMMAND } from './constants'
import { checkArguments } from "./helpers/argument-checker";
import { channelCreatorCfg } from "./arguments.cfg";
/**
 * Declares a channel for processed commands in the service interface
 * @param {Array} types Types of processed commands
 * @param {string} context Application context e.g. namespace of command
 * @returns {Function} Function for receiving commands from channel
 */
export default function execute(types: Array<string>, context: string): Function {
    checkArguments([types, context],'execute', channelCreatorCfg)
    const channel = channelCreator(types, context)
    return takeChannelCreator(TYPE_COMMAND, channel)
}
