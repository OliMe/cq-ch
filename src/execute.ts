import { channelCreator, takeChannelCreator } from './helpers/channel-creators';
import { TYPE_COMMAND } from './constants';
import { checkChannelCreator } from './helpers/argument-checkers';
import { Context, OutputMessage, Types } from './types';

/**
 * Declares a channel for processed commands in the service interface.
 * @param types Types of processed commands.
 * @param context Application context e.g. Namespace of command.
 * @return Function for receiving commands from channel.
 */
export default function execute(types: Types, context: Context) {
  checkChannelCreator('execute', types, context);
  const channel = channelCreator<OutputMessage<undefined>>(types, context);
  return takeChannelCreator<OutputMessage<undefined>>(TYPE_COMMAND, channel)();
}
