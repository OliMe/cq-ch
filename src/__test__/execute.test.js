import execute from '../execute'
import { TYPE_COMMAND } from '../constants'
import command from "../command";

describe('execute', () => {
    it('creates execute channel successfully', () => {
        expect(execute(['test_type'], 'test')).toBeInstanceOf(Function)
    })
    it('creates execute channel unsuccessfully', () => {
        const createChannelWithoutArguments = () => execute()
        const createChannelWithFirstIncorrectArgument = () => execute('not array')
        const createChannelWithFirstArgumentEmptyArray = () => execute([], 'test')
        const createChannelWithoutSecondArgument = () => execute(['test_type'])
        const createChannelWithSecondIncorrectArgument = () => execute(['test_type'], 200)
        expect(createChannelWithoutArguments).toThrowErrorMatchingSnapshot()
        expect(createChannelWithFirstIncorrectArgument).toThrowErrorMatchingSnapshot()
        expect(createChannelWithFirstArgumentEmptyArray).toThrowErrorMatchingSnapshot()
        expect(createChannelWithoutSecondArgument).toThrowErrorMatchingSnapshot()
        expect(createChannelWithSecondIncorrectArgument).toThrowErrorMatchingSnapshot()
    })
})
