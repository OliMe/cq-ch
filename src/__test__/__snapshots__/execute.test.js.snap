// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execute creates execute channel unsuccessfully 1`] = `"Failed to execute 'execute': first argument must be an Array with at least one element."`;

exports[`execute creates execute channel unsuccessfully 2`] = `"Failed to execute 'execute': first argument must be an Array with at least one element."`;

exports[`execute creates execute channel unsuccessfully 3`] = `"Failed to execute 'execute': first argument must be an Array with at least one element."`;

exports[`execute creates execute channel unsuccessfully 4`] = `"Failed to execute 'execute': second argument must be a string."`;

exports[`execute creates execute channel unsuccessfully 5`] = `"Failed to execute 'execute': second argument must be a string."`;
